pynest/nest/lib/hl_api_types.py

```diff
@@ -614,36 +617,72 @@ def __getitem__(self, key):
+            conns.print_full = True
+            print(conns)
         """
-        srcs = self.get('source')
-        trgt = self.get('target')
+        params = self.get()
+        srcs = params['source']
+        trgt = params['target']
+        wght = params['weight']
+        dlay = params['delay']
+        s_model = params['synapse_model']
 
         if isinstance(srcs, int):
             srcs = [srcs]
-        if isinstance(trgt, int):
             trgt = [trgt]
+            wght = [wght]
+            dlay = [dlay]
+            s_model = [s_model]
 
-        # 35 is arbitrarily chosen.
-        if len(srcs) < 35:
-            source = '| source | ' + ''.join(str(e)+', ' for e in srcs) + '|'
-            target = '| target | ' + ''.join(str(e)+', ' for e in trgt) + '|'
-        else:
-            source = ('| source | ' + ''.join(str(e)+', ' for e in srcs[:15]) +
-                      '... ' + ''.join(str(e)+', ' for e in srcs[-15:]) + '|')
-            target = ('| target | ' + ''.join(str(e)+', ' for e in trgt[:15]) +
-                      '... ' + ''.join(str(e)+', ' for e in trgt[-15:]) + '|')
+        src_h = 'source'
+        trg_h = 'target'
+        sm_h = 'synapse model'
+        w_h = 'weight'
+        d_h = 'delay'
+
+        max_char_sm = len(max(s_model, key=len))
 
-        borderline_s = '*--------*' + '-'*(len(source) - 12) + '-*'
-        borderline_t = '*--------*' + '-'*(len(target) - 12) + '-*'
-        borderline_m = max(borderline_s, borderline_t)
+        # Find maximum number of characters for each column, used to determine width of column
+        src_len = len(src_h) + 2 if floor(log(max(srcs), 10)) <= len(src_h) + 2 else floor(log(max(srcs), 10))
+        trg_len = len(trg_h) + 2 if floor(log(max(trgt), 10)) <= len(trg_h) + 2 else floor(log(max(trgt), 10))
+        sm_len = len(sm_h) + 2 if max_char_sm <= len(sm_h) + 2 else max_char_sm
+        w_len = len(w_h) + 2
+        d_len = len(d_h) + 2
+
+        # 35 is arbitrarily chosen.
+        if len(srcs) >= 35 and not self.print_full:
+            srcs = srcs[:15] + [u'\u22EE '] + srcs[-15:]
+            trgt = trgt[:15] + [u'\u22EE '] + trgt[-15:]
+            wght = wght[:15] + [u'\u22EE '] + wght[-15:]
+            dlay = dlay[:15] + [u'\u22EE '] + dlay[-15:]
+            s_model = s_model[:15] + [u'\u22EE '] + s_model[-15:]
+
+        headers = f'{src_h:^{src_len}} {trg_h:^{trg_len}} {sm_h:^{sm_len}} {w_h:^{w_len}} {d_h:^{d_len}}' + '\n'
+        boarders = '-'*src_len + ' ' + '-'*trg_len + ' ' + '-'*sm_len + ' ' + '-'*w_len + ' ' + '-'*d_len + '\n'
+        output = ''.join('{:>{}}'.format(str(s), src_len-1) +
+                         ' {:>{}}'.format(str(t), trg_len) +  # need to indent because of space between columns
+                         ' {:>{}}'.format(str(sm), sm_len) +
+                         ' {:>{num}.{deci}}'.format(w, num=w_len, deci=4) +
+                         ' {:>{num}.{deci}}'.format(d, num=d_len, deci=4) +
```

Comment:
The following should fix the "missing leading zero" problem and give exponential notation automatically:
```suggestion
                         ' {:>#{num}.{deci}g}'.format(w, num=w_len, deci=4) +
                         ' {:>#{num}.{deci}g}'.format(d, num=d_len, deci=4) +
```

Reply:
This will not work when we have many connections, because then we insert the unicode for vertical ellipsis (see [here](https://github.com/nest/nest-simulator/pull/1836/files#diff-bdf50826d5c778982ced8b68c069efde8bda3349c0a88e4888cd14346e42f769R672-R673)), and both `#` and `g` only works on numbers. We therefore get the following error when we print a SynapseCollection with many connections: `ValueError: Unknown format code 'g' for object of type 'str'` (If we do not use `g` we still get the error, but with `#` instead)

Reply:
How about a helper such as
```python
def fmt_row_(s, t, sm, w, d, sl, tl, sml, wn, wd, dn, dd):
    try:
        return f'{s:>{sl}d} {t:>{tl}d} {sm:>{sml}s} {w:>#{wn}.{wd}g} {d:>#{dn}.{dd}g}'
    except ValueError:
        return f'{s:>{sl}} {t:>{tl}} {sm:>{sml}} {w:>{wn}.{wd}} {d:>{dn}.{dd}}'
```
and then just
```python
'\n'.join(fmt_row_(s, t, sm, w, d, ...) 
           for s, t, sm, w, d in zip(srcs, trgt, s_model, wght, dlay))
```
If you define `fmt_row_()` locally in the method, I think you don't even need to pass the length parameters.

Reply:
Smart! I will take a look.